File: packages/shared/onboarding/content/exercises.ts
import { ExerciseSchema, type Exercise } from "./schema";

const allPaths = ["business", "technical", "compliance", "admin"] as const;

// The unowned-process SPARQL ASK CE-03/CE-03b both verify against (PRD
// E4-S1 exercise table: "Process nodes with no performedBy Actor").
// `weave:label`, not `rdfs:label` -- ADR-005 (see rdf/patterns.py).
const UNOWNED_PROCESS_ASK =
  "PREFIX weave: <https://weave.io/ontology/>\n" +
  "ASK { GRAPH ?g { ?p a weave:Process . FILTER NOT EXISTS { ?p weave:performedBy ?a } } }";

const ce01: Exercise = {
  exerciseId: "CE-01",
  paths: [...allPaths],
  phase: "m1",
  goalKey: "onboarding.exercise.ce-01.goal",
  stepKeys: ["onboarding.exercise.ce-01.step1", "onboarding.exercise.ce-01.step2", "onboarding.exercise.ce-01.step3"],
  // Both the entity-list and missing-property views must be visited in one
  // session (Implementation Hints) -- the one deliberately soft/client-signal
  // check (PRD marks CE-01 a UI-nav signal), kept honest by requiring both.
  completion: { kind: "nav_signal", signal: "entity-list-viewed,missing-property-viewed" },
};

/** Verified server-side by static ASK over the sandbox graph, after the
 * NL write commits via CE-WRITE-1 (AC-009-02/03). */
const ce02: Exercise = {
  exerciseId: "CE-02",
  paths: [...allPaths],
  phase: "m1",
  goalKey: "onboarding.exercise.ce-02.goal",
  stepKeys: ["onboarding.exercise.ce-02.step1", "onboarding.exercise.ce-02.step2", "onboarding.exercise.ce-02.step3"],
  completion: {
    kind: "sparql_ask",
    ask:
      "PREFIX weave: <https://weave.io/ontology/>\n" +
      'ASK { GRAPH ?g { ?s a weave:Class ; weave:label "Outdoor Furniture" } }',
  },
};

/** Technical-only (FR-016). */
const ce03: Exercise = {
  exerciseId: "CE-03",
  paths: ["technical"],
  phase: "m1",
  goalKey: "onboarding.exercise.ce-03.goal",
  stepKeys: ["onboarding.exercise.ce-03.step1", "onboarding.exercise.ce-03.step2", "onboarding.exercise.ce-03.step3"],
  completion: { kind: "sparql_ask", ask: UNOWNED_PROCESS_ASK },
};

/** Business (FR-016) -- the guided NL equivalent of CE-03's raw-SPARQL
 * exercise; same underlying fact, same ASK (AC-009-06). */
const ce03b: Exercise = {
  exerciseId: "CE-03b",
  paths: ["business", "compliance", "admin"],
  phase: "m1",
  goalKey: "onboarding.exercise.ce-03b.goal",
  stepKeys: ["onboarding.exercise.ce-03b.step1", "onboarding.exercise.ce-03b.step2", "onboarding.exercise.ce-03b.step3"],
  completion: { kind: "sparql_ask", ask: UNOWNED_PROCESS_ASK },
};

const ge01: Exercise = {
  exerciseId: "GE-01",
  paths: [...allPaths],
  phase: "m1",
  goalKey: "onboarding.exercise.ge-01.goal",
  stepKeys: ["onboarding.exercise.ge-01.step1", "onboarding.exercise.ge-01.step2", "onboarding.exercise.ge-01.step3"],
  completion: { kind: "canvas_state", state: "spotlight-active" },
};

const ge02: Exercise = {
  exerciseId: "GE-02",
  paths: [...allPaths],
  phase: "m1",
  goalKey: "onboarding.exercise.ge-02.goal",
  stepKeys: ["onboarding.exercise.ge-02.step1", "onboarding.exercise.ge-02.step2", "onboarding.exercise.ge-02.step3"],
  completion: { kind: "canvas_state", state: "heatmap-overlay-active" },
};

export const EXERCISES: Exercise[] = [ce01, ce02, ce03, ce03b, ge01, ge02].map((e) => ExerciseSchema.parse(e));

File: packages/backend/src/weave_backend/onboarding/exercises.py
"""TASK-009: server-side mirror of `packages/shared/onboarding/content/
exercises.ts` (TASK-003). Kept as a small static registry, not a codegen
step, per ADR-006's "ASK strings are static config, PR-reviewed" -- the
backend needs the *completion* shape (to dispatch the check) and the
*gating* shape (paths/variant) server-side too, since the frontend's own
gating is UI-only and must never be the sole enforcement (AC-009-01/02).

Keep this in sync with `exercises.ts` by hand when either changes -- six
entries, reviewed in the same PR, is cheaper than a cross-language codegen
pipeline for M1.
"""

from __future__ import annotations

from typing import Literal, TypedDict

RolePath = Literal["business", "technical", "compliance", "admin"]

CompletionKind = Literal["sparql_ask", "canvas_state", "nav_signal", "write_commit"]


class Completion(TypedDict, total=False):
    kind: CompletionKind
    ask: str
    state: str
    signal: str


class ExerciseDef(TypedDict):
    paths: tuple[RolePath, ...]
    completion: Completion


_ALL_PATHS: tuple[RolePath, ...] = ("business", "technical", "compliance", "admin")

_UNOWNED_PROCESS_ASK = (
    "PREFIX weave: <https://weave.io/ontology/>\n"
    "ASK { GRAPH ?g { ?p a weave:Process . FILTER NOT EXISTS { ?p weave:performedBy ?a } } }"
)

EXERCISES: dict[str, ExerciseDef] = {
    "CE-01": {
        "paths": _ALL_PATHS,
        "completion": {
            "kind": "nav_signal",
            "signal": "entity-list-viewed,missing-property-viewed",
        },
    },
    "CE-02": {
        "paths": _ALL_PATHS,
        "completion": {
            "kind": "sparql_ask",
            "ask": (
                "PREFIX weave: <https://weave.io/ontology/>\n"
                'ASK { GRAPH ?g { ?s a weave:Class ; weave:label "Outdoor Furniture" } }'
            ),
        },
    },
    "CE-03": {
        "paths": ("technical",),
        "completion": {"kind": "sparql_ask", "ask": _UNOWNED_PROCESS_ASK},
    },
    "CE-03b": {
        "paths": ("business", "compliance", "admin"),
        "completion": {"kind": "sparql_ask", "ask": _UNOWNED_PROCESS_ASK},
    },
    "GE-01": {
        "paths": _ALL_PATHS,
        "completion": {"kind": "canvas_state", "state": "spotlight-active"},
    },
    "GE-02": {
        "paths": _ALL_PATHS,
        "completion": {"kind": "canvas_state", "state": "heatmap-overlay-active"},
    },
}

#: AC-009-01: the read-only sandbox variant locks exercises that write --
#: only CE-02 (add a category) writes; every other M1 exercise is read-only.
WRITE_EXERCISE_IDS = frozenset({"CE-02"})


class GateResult:
    """Why an exercise is or isn't available to this caller right now."""

    def __init__(self, *, available: bool, reason: str | None = None) -> None:
        self.available = available
        self.reason = reason


def gate_exercise(
    exercise_id: str, *, role_path: RolePath, path_variant: Literal["default", "read_only"]
) -> GateResult:
    """AC-009-01/06: server-side gating -- the frontend's own gating is UI
    sugar, never the enforcement (this function is what `check_exercise_
    route` actually trusts).
    """
    exercise = EXERCISES.get(exercise_id)
    if exercise is None:
        return GateResult(available=False, reason="unknown_exercise")
    if role_path not in exercise["paths"]:
        return GateResult(available=False, reason="path_gated")
    if path_variant == "read_only" and exercise_id in WRITE_EXERCISE_IDS:
        return GateResult(available=False, reason="read_only_locked")
    return GateResult(available=True)


def available_exercises(
    *, role_path: RolePath, path_variant: Literal["default", "read_only"]
) -> list[str]:
    """T8: exercise ids `gate_exercise` currently allows this caller to
    check -- `GET /onboarding/state` exposes this so the checklist client
    can skip an exercise it can't complete, instead of POSTing to `/check`
    and taking a 403 (`path_gated` / `read_only_locked`) it could have
    avoided.
    """
    return [
        exercise_id
        for exercise_id in EXERCISES
        if gate_exercise(exercise_id, role_path=role_path, path_variant=path_variant).available
    ]
